### ISSUE TITLE
WebhookParser does not properly parse email addresses containing a comma

### ISSUE THREAD

**Issue by alex21777** [Jericho/StrongGrid#185]
in WebhookParsers class 
ParseEmailAddresses(string rawEmailAddresses)

please replace 
var rawEmails = rawEmailAddresses.Split(new[] { ',' }, StringSplitOptions.RemoveEmptyEntries);

with 
var rawEmails = Regex.Split(rawEmailAddresses, ",(?=(?:[^']*'[^']*')*[^']*$)");

Reason:
recipient name might contain commas.
For example:
Smith, John<john@foo.com>

Thank you,
Alex

**Comment by Jericho** [comment]
The intent of the `ParseEmailAddresses` is to parse a string into an array of email addresses. So, something like: `bob <bob@example.com>, tom <tom@example.com>`. That's why I decided to split on the comma.

You're right that my logic does not handle recipient names containing a comma. However, your [regex solution](http://regexstorm.net/tester?p=%2c%28%3f%3d%28%3f%3a%5b%5e%27%5d*%27%5b%5e%27%5d*%27%29*%5b%5e%27%5d*%24%29&i=Smith%2c+John%3cjohn%40foo.com%3e) doesn't seem to solve the problem either.

**Comment by Jericho** [comment]
Also, do you have an example of a string containing more than one email address where one of the addresses contain a comma? I'm guessing it will look something like `"Smith, John <john@foo.com>", "Bob <bob@example.com>"`. My point is that there will be some way to distinguish between the comma in the recipient's name and the comma separating multiple recipients.

**Comment by alex21777** [comment]
I gave you a solution – see below

var rawEmails = Regex.Split(rawEmailAddresses, ",(?=(?:[^']*'[^']*')*[^']*$)");

**Comment by Jericho** [comment]
Thank you for the solution. Much appreciated. However, before I accept it I want to validate that it indeed solves the problem you are reporting. That's why I asked you to help me by providing a few examples. In particular, I would like to see an example of a string containing multiple recipient addresses where one of the addresses contain a comma.

I used an online tool called regexstorm to validate the one example you have already provided and I have observed that [your proposed solution](http://regexstorm.net/tester?p=%2c%28%3f%3d%28%3f%3a%5b%5e%27%5d*%27%5b%5e%27%5d*%27%29*%5b%5e%27%5d*%24%29&i=Smith%2c+John%3cjohn%40foo.com%3e) does not solve the simple case `Smith, John<john@foo.com>` that you have presented.

**Comment by alex21777** [comment]
The following is an example:

"Smith, John”<john@foo.com>, “Bob”<bob@example.com>

**Comment by Jericho** [comment]
Thank you for this sample. Unfortunately, it doesn't look like your proposed regex solves this case either.

Have a look at [this](http://regexstorm.net/tester?p=%2c%28%3f%3d%28%3f%3a%5b%5e%27%5d*%27%5b%5e%27%5d*%27%29*%5b%5e%27%5d*%24%29&i=Smith%2c+John%3cjohn%40foo.com%3e). You'll see that your proposed solution parses `"Smith, John”<john@foo.com>, “Bob”<bob@example.com>` into three string which is not the desired behavior.

**Comment by alex21777** [comment]
Not sure about your tester.

It works when you run the real code.

**Comment by Jericho** [comment]
I wrote some unit tests to validate your proposed regex and all three tests fail. You said that your proposed solution "works when you run the real code". Can you show me the code you wrote to validate your solution?

Here are the unit tests I wrote:

```
using Shouldly;
using System.Text.RegularExpressions;
using Xunit;

namespace StrongGrid.UnitTests
{
    public class SplitEmailAddressesTests
    {
        private const string REGEX_PATTERN = ",(?=(?:[^']*'[^']*')*[^']*$)";

        [Fact]
        public void Split_multiple_email_addresses()
        {
            // Arrange
            var input = "\"Smith, Bob\" <bob@example.com>, someone_else@example.com";

            // Act
            var parts = Regex.Split(input, REGEX_PATTERN);

            // Assert
            parts.Length.ShouldBe(2);
            parts[0].ShouldBe("\"Smith, Bob\" <bob@example.com>");
            parts[1].ShouldBe("someone_else@example.com");
        }

        [Fact]
        public void Split_single_email_addresses_with_comma()
        {
            // Arrange
            var input = "\"Smith, Bob\" <bob@example.com>";

            // Act
            var parts = Regex.Split(input, REGEX_PATTERN);

            // Assert
            parts.Length.ShouldBe(1);
            parts[0].ShouldBe("\"Smith, Bob\" <bob@example.com>");
        }

        [Fact]
        public void Split_single_email_addresses_without_comma()
        {
            // Arrange
            var input = "someone_else@example.com";

            // Act
            var parts = Regex.Split(input, REGEX_PATTERN);

            // Assert
            parts.Length.ShouldBe(1);
            parts[0].ShouldBe("someone_else@example.com");
        }
    }
}
```

**Comment by alex21777** [comment]
My apologies.

Sendgrid surrounds name with single quotes (not double quotes) like the following:

'Smith, John'<john@foo.com>,foo@foo.com

Thank you,
Alex

**Comment by Jericho** [comment]
Oh I see. Your Regex is looking for single quote.

Are you sure that SendGrid uses quotes (instead of double-quotes)? I was under the impression they used double quotes. I will run a few tests to verify.

**Comment by Jericho** [comment]
I configured my Inbound Parse Webhook to post to http://requests.in for easy visualisation and, as you can see, the name Bob Smith is surrounded by double quotes. 

But maybe you experience is different? Maybe there are some scenarios where SendGrid uses quotes?

![img_0157](https://user-images.githubusercontent.com/112710/29799544-48789340-8c32-11e7-9506-39d3a292987c.PNG)

**Comment by Jericho** [comment]
Hmmmmm, that's intriguing. I use Outlook as well.

**Comment by Jericho** [comment]
Is there any other info you can give me to reproduce the issue where Sendgrid uses quotes? Is this something you can reproduce consistently?

**Comment by Jericho** [comment]
@alex21777 I have continued my investigation but I can't find any evidence of the single-quote ever being used as you said and I can't reproduce either. 

I found RFC2822 which describes the format of email messages [mentions specifically that ASCII character 34](https://tools.ietf.org/html/rfc2822#section-3.2.5) (which is the double-quote) must be used to surround quoted strings.

I also found [Microsoft's own email address parser](https://referencesource.microsoft.com/#system/net/System/Net/mail/MailAddressParser.cs
) only checks for [double quote](https://referencesource.microsoft.com/#system/net/System/Net/mail/MailBnfHelper.cs,b11a50e944f52585).

So I really need your help to guide on how to reproduce this situation with the single-quote.

Having said all this, the fact remains that the parsing logic in the StrongGrid Inbound parser is flawed (as you pointed out) and does not handle commas in the "name". I continue my research to figure out the best solution.
